sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala

```diff
@@ -17,53 +17,234 @@
+  private[sql] val ordering =
+    (x: Float, y: Float) => SQLOrderingUtil.compareFloats(x, y)
+}
+case object PhysicalFloatType extends PhysicalFloatType with PhysicalPrimitiveType
+
+class PhysicalIntegerType() extends PhysicalIntegralType {
+  // The companion object and this class is separated so the companion object also subclasses
+  // this type. Otherwise, the companion object would be of type "IntegerType$" in byte code.
+  // Defined with a private constructor so the companion object is the only possible instantiation.
+  private[sql] type InternalType = Int
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+  private[sql] val ordering = implicitly[Ordering[InternalType]]
+}
+case object PhysicalIntegerType extends PhysicalIntegerType with PhysicalPrimitiveType
+
+class PhysicalLongType() extends PhysicalIntegralType {
+  // The companion object and this class is separated so the companion object also subclasses
+  // this type. Otherwise, the companion object would be of type "LongType$" in byte code.
+  // Defined with a private constructor so the companion object is the only possible instantiation.
+  private[sql] type InternalType = Long
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+  private[sql] val ordering = implicitly[Ordering[InternalType]]
+}
+case object PhysicalLongType extends PhysicalLongType with PhysicalPrimitiveType
 
 case class PhysicalMapType(keyType: DataType, valueType: DataType, valueContainsNull: Boolean)
     extends PhysicalDataType
 
-class PhysicalNullType() extends PhysicalPrimitiveType
-case object PhysicalNullType extends PhysicalNullType
-
-class PhysicalShortType() extends PhysicalPrimitiveType
-case object PhysicalShortType extends PhysicalShortType
-
-class PhysicalStringType() extends PhysicalDataType
+class PhysicalNullType() extends PhysicalDataType
+case object PhysicalNullType extends PhysicalNullType with PhysicalPrimitiveType
+
+class PhysicalShortType() extends PhysicalIntegralType {
+  private[sql] type InternalType = Short
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+  private[sql] val ordering = implicitly[Ordering[InternalType]]
+}
+case object PhysicalShortType extends PhysicalShortType with PhysicalPrimitiveType
+
+class PhysicalStringType() extends PhysicalAtomicType {
+  // The companion object and this class is separated so the companion object also subclasses
+  // this type. Otherwise, the companion object would be of type "StringType$" in byte code.
+  // Defined with a private constructor so the companion object is the only possible instantiation.
+  private[sql] type InternalType = UTF8String
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+  private[sql] val ordering = implicitly[Ordering[InternalType]]
+}
 case object PhysicalStringType extends PhysicalStringType
 
 case class PhysicalStructType(fields: Array[StructField]) extends PhysicalDataType
```

Comment:
This is orderable right?

Reply:
`case class StructType(fields: Array[StructField]) extends DataType with Seq[StructField] {`

it does not extend the type that offers ordering?

Reply:
but it is orderable :)